cumulusci/tasks/bulkdata/data_generation/data_generator_runtime.py

```diff
@@ -0,0 +1,236 @@
+        assert isinstance(definition, str), definition
+        if "<<" in definition or "<%" in definition:
+            try:
+                template = self.template_compiler.from_string(definition)
+                return DynamicEvaluator(template)
+            except jinja2.exceptions.TemplateSyntaxError as e:
+                raise DataGenSyntaxError(str(e), None, None) from e
+        else:
+            return lambda RuntimeRuntimeContext: definition
+
+
+class RuntimeContext:
+    """Current context object. RuntimeContext objects form a linked list-based stack"""
+
+    current_id = None
+    obj: Optional["ObjectRow"] = None
+    today = date.today()
+    template_evaluator_factory = JinjaTemplateEvaluatorFactory()
+    counter = 0
+
+    def __init__(
+        self,
+        parent: Optional["RuntimeContext"],
+        current_table_name: Optional[str],
+        output_stream=None,
+        options=None,
+    ):
+        self.parent = parent
+        self.current_table_name = current_table_name
+        options = options or {}
+        self.field_values: Dict[str, Any] = {}
+
+        if parent:
+            self.globals: Globals = parent.globals
+            self.output_stream: Any = parent.output_stream
+            self.options: Dict = {**parent.options}
+        else:  # root RuntimeContext
+            self.globals = Globals()
+            self.output_stream = output_stream
+            self.options = {**options}
+
+    def incr(self):
+        """Increments the local counter for an object type"""
+        self.counter += 1
+
+    def generate_id(self):
+        self.current_id = self.globals.id_manager.generate_id(self.current_table_name)
+        return self.current_id
+
+    def register_object(self, obj, name=None):
+        self.obj = obj
+        self.globals.register_object(obj, name)
+
+    def register_intertable_reference(self, table_name_from, table_name_to, fieldname):
+        self.globals.register_intertable_reference(
+            table_name_from, table_name_to, fieldname
+        )
+
+    def register_field(self, field_name, field_value):
+        """Register each field value to be ready to inject them into template"""
+        self.field_values[field_name] = field_value
+
+    def field_vars(self):
+        return {
+            "id": self.current_id,
+            "this": self.obj,
+            "today": self.today,
+            "fake": faker_template_library,
+            "fake_i18n": lambda locale: FakerTemplateLibrary(locale),
+            "number": self.counter,
+            **self.options,
+            **self.globals.object_names,
+            **self.field_values,
+        }
+
+    def field_funcs(self):
+        def curry(func):
+            rc = partial(func, self)
+            if hasattr(func, "lazy"):
+                rc.lazy = func.lazy
+            return rc
+
+        funcs = {name: curry(func) for name, func in template_funcs.items()}
+        return {**funcs}
```

Comment:
It looks like `field_funcs` gets called once for each SimpleValue and StructuredValue, but could be memoized so that it's only done once per context.